Fix IndexOutOfRangeException when an added/removed attribute spans a buffer size boundary

SteveSandersonMS opened the pull request:
**NOTE: I'm not actually proposing this for servicing right now, because it doesn't look like there would be a major benefit from including it in RC1, and including it in RC2 would be perfectly sufficient. However I'm writing up the ask mode template anyway just in case people disagree.**

Extremely rarely, component rendering could cause IndexOutOfRangeException.

## Description

If a component rendered a very specific number of frames, and then updated to render exactly one more frame, and that new frame was an attribute, and it was the last frame in the component, the diffing system would throw `IndexOutOfRangeException`. This was due to a bug in checking whether there are more attributes to process.

Fixes #49192

## Customer Impact

Fairly low, because it's really hard to reproduce this, even if you're doing so on purpose. However, total correctness of the Blazor renderer is a high priority for us, and we do not want any edge cases where it can fail.

In the event that the issue does occur, it would behave like an unhandled exception from application code:

 * For Blazor WebAssembly, the yellow error bar would appear, but the app would continue running
 * For Blazor Server, the current user's circuit would be terminated, but other server activity would be unaffected
 * For Blazor WebView, the unhandled exception callback would run

## Regression?

- [ ] Yes
- [x] No

[If yes, specify the version the behavior has regressed from]

## Risk

- [ ] High
- [ ] Medium
- [x] Low

Low because the code change is small enough (4 lines) to reason clearly that it only makes a difference in the case where it would have failed before. The updated boundary condition is simple to understand.

## Verification

- [x] Manual (required)
- [x] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

SteveSandersonMS committed:
Fix IndexOutOfRangeException when an added/removed attribute spans a buffer size boundary

SteveSandersonMS commented on src/Components/Components/test/RenderTreeDiffBuilderTest.cs:
```diff
@@ -2319,6 +2319,73 @@ public void RecognizesNamedEventChangingAssignedName()
+                Assert.Equal("myattribute_final", entry.RemovedAttributeName);
+            });
+    }
+
```
Both of these new unit tests do fail with `IndexOutOfRangeException` without the fix, and of course pass with the fix.

campersau commented on src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs:
```diff
@@ -18,6 +18,8 @@ enum DiffAction { Match, Insert, Delete }
     // hit the "old < new" code path during diffing so we only have to check for it in one place.
     public const int SystemAddedAttributeSequenceNumber = int.MinValue;
 
+    private static RenderTreeFrame NoSuchAttributeFrame = RenderTreeFrame.Attribute(int.MaxValue, null, null);
```
```suggestion
    private static readonly RenderTreeFrame NoSuchAttributeFrame = RenderTreeFrame.Attribute(int.MaxValue, null, null);
```

SteveSandersonMS committed:
Update src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs

Co-authored-by: campersau <buchholz.bastian@googlemail.com>

SteveSandersonMS commented on src/Components/Components/src/RenderTree/RenderTreeDiffBuilder.cs:
```diff
@@ -18,6 +18,8 @@ enum DiffAction { Match, Insert, Delete }
     // hit the "old < new" code path during diffing so we only have to check for it in one place.
     public const int SystemAddedAttributeSequenceNumber = int.MinValue;
 
+    private static RenderTreeFrame NoSuchAttributeFrame = RenderTreeFrame.Attribute(int.MaxValue, null, null);
```
Good point - have made it readonly now.

SteveSandersonMS replied:
Actually I've changed it further to avoid having `NoSuchAttributeFrame` at all. If two of us on the team can't immediately tell whether that was going to cause struct copying, it's not a language feature we should be relying on.

SteveSandersonMS committed:
Avoid using obscure C# language features we can't easily reason about

There's nothing in the syntax to indicate whether NoSuchAttributeFrame would be copied or not
